packages/vat-data/src/vat-data-bindings.js

```diff
@@ -91,19 +93,115 @@ export const partialAssign = (target, source) => {
+ */
+const collisions = new WeakMap();
+
+/**
+ * HAZARD: `assertUniqueProvide` uses top-level mutable state, but only for
+ * diagnostic purposes.
+ * Aside from being able to provoke it to throw an error, it is not
+ * otherwise usable as a top-level communications channel. The throwing-or-not
+ * of the error is the only way in which this top level state is observable.
+ *
+ * TODO since this still does violate our security assumptions, we should
+ * consider making it development-time only, and turn it off in production.
+ * Or better, adapt it to stop using top-level state.
+ *
+ * The condition it checks for is that, within one incarnation, the
+ * `provide` is never called twice with the same `mapStore`
+ * and the same `key`. For a given `mapStore` and `key`, there should be at
+ * most one `provide` call but any number of `get` calls.
+ *
+ * When the `mapStore` is baggage, for purposes of upgrade, then we
+ * expect `provide` to be called with the same `baggage` and `key`
+ * (and some `makeValue`) in each incarnation of that vat.
+ * In the incarnation that first registers this `baggage`,`key` pair, the
+ * `provide` will actually call the `makeValue`. The one `provide` call
+ * for this pair in each succeeding incarnation will merely retrieve the value it
+ * inherited from the previous incarnation. But even for this case, it should
+ * only call `provide` for that pair once during that incarnation.
+ *
+ * @param {import('./types.js').Baggage} baggage
+ * @param {string} key
+ * @param {(key?: string) => unknown} makeValue
+ */
+const assertUniqueProvide = (baggage, key, makeValue) => {
+  if (collisions.has(baggage)) {
+    const submap = /** @type {Map<string, (k?:string) => any>} */ (
+      collisions.get(baggage)
+    );
+    if (submap.has(key)) {
+      if (submap.get(key) === makeValue) {
+        // The error message also tells us if the colliding calls happened
+        // to use the same `makeValue` only as a clue about whether
+        // these are two separate evaluations at of the same call site or
+        // two distinct call sites. However, this clue is unreliable in
+        // both directions.
+        assert.fail(X`provide collision on ${q(key)} with same makeValue`);
+      } else {
+        assert.fail(X`provide collision on ${q(key)} with different makeValue`);
+      }
+    } else {
+      submap.set(key, makeValue);
+    }
+  } else {
+    const submap = /** @type {Map<string, (k?:string) => any>} */ (
+      new Map([[key, makeValue]])
+    );
+    collisions.set(baggage, submap);
+  }
+};
+
+/**
+ * Unlike `provide`, `provideOnce` can be called at most once
+ * within any vat incarnation with a given `baggage`,`key` pair.
+ *
+ * `provideOnce` should only to be used to populate baggage,
+ * where the total number of calls to `provideOnce` must be
+ * low cardinality, since we keep the bookkeeping to detect collisions
+ * in normal language-heap memory. All the other baggage-oriented
```

Comment:
When this is configured to be inert in production, this warning will no longer apply, correct?

What would it take to make this track only in dev mode? I think the plan for that should be sketched before merging, so we know what to design this API towards.

Reply:
Ah. How about I do condition it on a development time switch that can be sensed when running in unit testing, but cannot be sensed even in development from vats under swingset? This would 
   * give us some benefit,
   * would enable us to test the
   * would fail-safe under current SwingSet, failing to catch errors during some development, rather than possibly violating a security property in production.

With this plan, I am putting the PR back into Ready for Review, but will not close this conversation until it is safe by that plan.

Reply:
Leaving this conversation open until I add a tests of the mechanism itself, conditional on the presence or absence of `process.env.DETECT_PROVIDE_COLLISIONS` (the same option setting that an environment variable would produce).